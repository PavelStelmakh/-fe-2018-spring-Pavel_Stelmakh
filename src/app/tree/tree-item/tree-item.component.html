<div (click)="expandTree()">
    <div><fa-icon [icon]="expandItem ? faCaretDown : faCaretLeft" size="1x" style="color: #ccc"></fa-icon></div>
  <span>{{itemTree}}</span>
</div>
<div *ngIf="expandItem">
  <app-tree-item *ngFor="let item of items" [item]="item"></app-tree-item>
</div>