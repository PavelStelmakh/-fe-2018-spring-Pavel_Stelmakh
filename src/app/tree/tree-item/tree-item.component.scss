:host {
    text-align: left;

    div > div {
        width: 10px;
        display: inline-block;
    }

    div + div {
        margin-left: 10px;
    }
}