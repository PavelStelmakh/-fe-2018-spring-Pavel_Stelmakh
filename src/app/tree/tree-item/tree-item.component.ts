import { 
  Component,
  Input,
  OnInit,
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  ViewChildren,
  QueryList 
} from '@angular/core';
import { TreeService } from '../tree.service';
import { faCaretDown, IconDefinition, faCaretLeft } from '@fortawesome/free-solid-svg-icons';

@Component({
  selector: 'app-tree-item',
  templateUrl: './tree-item.component.html',
  styleUrls: ['./tree-item.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class TreeItemComponent implements OnInit {
  items: string[];
  expandItem: boolean;
  @Input('item') itemTree: string;
  faCaretDown: IconDefinition = faCaretDown;
  faCaretLeft: IconDefinition = faCaretLeft;
  private path: string[];
  @ViewChildren(TreeItemComponent) itemComponents: QueryList<TreeItemComponent>;

  constructor(public treeService: TreeService, private cdr: ChangeDetectorRef) { }

  ngOnInit() {
    this.path = this.treeService.path.slice();
    this.treeService.search.subscribe((path: string) => {
      this.filter(path);
      this.cdr.markForCheck();
    });
    this.items = [];
    this.expandItem = false;
  }

  expandTree() {
    this.expandItem = !this.expandItem;
    if (this.expandItem) {
      if (this.treeService.path[this.path.length]) {
        this.treeService.setPath(this.path.length, this.itemTree);
      } else {
        this.treeService.pushPath(this.itemTree);
      }
      this.items = this.treeService.getTreeItems();
    } else {
      this.treeService.popPath(this.path);
    }
  }

  filter(Searchingitem: string) {
    this.items = this.itemComponents
    .filter((component: TreeItemComponent) => Searchingitem === component.itemTree || component.expandItem)
    .map((component: TreeItemComponent) => component.itemTree);
  }

}
