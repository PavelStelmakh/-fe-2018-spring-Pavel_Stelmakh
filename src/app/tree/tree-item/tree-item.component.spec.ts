import { TestBed, async, ComponentFixture } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { TreeItemComponent } from './tree-item.component';

describe('TreeItemComponent', () => {
  let fixture: ComponentFixture<TreeItemComponent>;
  let component: TreeItemComponent;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        TreeItemComponent
      ],
      schemas: [
       NO_ERRORS_SCHEMA
      ]
    }).compileComponents();

  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TreeItemComponent);
    component = fixture.componentInstance;
  });

  it('renders markup to snapshot', () => {
    component.expandItem = true;
    fixture.detectChanges();
    
    expect(fixture).toMatchSnapshot();
  });

});
