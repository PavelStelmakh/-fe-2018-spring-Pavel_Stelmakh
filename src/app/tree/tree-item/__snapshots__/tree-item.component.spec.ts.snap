// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeItemComponent renders markup to snapshot 1`] = `
<app-tree-item
  cdr={[Function ViewRef_]}
  expandItem="false"
  faCaretDown={[Function Object]}
  faCaretLeft={[Function Object]}
  itemComponents={[Function QueryList]}
  items={[Function Array]}
  path={[Function Array]}
  treeService={[Function TreeService]}
>
  <div>
    <div>
      <fa-icon
        size="1x"
        style="color: #ccc"
      />
    </div>
    <span>
      
    </span>
  </div>
</app-tree-item>
`;
